contrib/win32/win32compat/win32_dirent.c

```diff
@@ -11,33 +11,92 @@
 
 #include "inc\dirent.h"
 #include "inc\libgen.h"
+#include "inc\defs.h"
 #include "misc_internal.h"
 
 
 struct DIR_ {
 	intptr_t hFile;
 	struct _wfinddata_t c_file;
 	int first;
+	wchar_t * nextdisk;
 };
 
+#define ATTR_ROOTDIR  UINT_MAX
+
+DIR * openrootdir(const char *name)
+{
+	int hr = 0;
+	DWORD dw;
+	DIR * pdir;
+	struct _wfinddata_t c_file = {0};
+	wchar_t * p;
+
+	dw = GetLogicalDriveStringsW(_ARRAYSIZE(c_file.name) - 2, c_file.name);
+	if (!dw) {
+		errno = ENODEV;
+		return NULL;
+	}
+	c_file.attrib = ATTR_ROOTDIR;
+	c_file.size = 0;
+	p = c_file.name;
+	while (*p) {
+		size_t len = wcslen(p);
+		if (len == 0)
+			break;
+		p += len + 1;
+		c_file.size++;
+	}
+	if (c_file.size == 0) {
+		errno = ENODEV;
+		return NULL;
+	}
+	pdir = malloc(sizeof(DIR));
+	if (!pdir) {
+		errno = ENOMEM;
+		return NULL;
+	}
+	memset(pdir, 0, sizeof(DIR));
+	pdir->hFile = 0;
+	memcpy(&pdir->c_file, &c_file, sizeof(c_file));
+	pdir->first = 1;
+
+	return pdir;
+}
+
 /* Open a directory stream on NAME.
    Return a DIR stream on the directory, or NULL if it could not be opened.  */
 DIR * opendir(const char *name)
 {
 	struct _wfinddata_t c_file;
 	intptr_t hFile;
 	DIR *pdir;
-	wchar_t searchstr[MAX_PATH];
+	wchar_t searchstr[PATH_MAX + 16];
```

Comment:
can you please explain me what is the issue with the existing code 
swprintf_s(searchstr, MAX_PATH, L"%s\\*.*", wname);